handsanitizer/FunctionCallerGenerator.cpp

```diff
@@ -0,0 +1,223 @@
+        std::stringstream output;
+        output << "typedef struct " << type->getCTypeName() << " {" << std::endl;
+        for (auto &member: type->getNamedMembers()) {
+            if (member.getType()->isArrayTy()) {
+                output << member.getType()->getArrayElementType()->getCTypeName() << " " << member.getName();
+                output << "[" << member.getType()->getArrayNumElements() << "]";
+            } else {
+                output << member.getType()->getCTypeName() << " " << member.getName();
+            }
+            output << ";" << std::endl;
+        }
+        output << "} " << type->getCTypeName() << ";" << std::endl;
+        return output.str();
+    }
+
+
+    std::string FunctionCallerGenerator::getUntypedArgumentNames(Function &f) {
+        std::stringstream output;
+        for (auto &args : f.arguments) {
+            output << args.getName() << ",";
+        }
+        auto retstring = output.str();
+        if (!retstring.empty())
+            retstring.pop_back(); //remove last ,
+
+        return retstring;
+    }
+
+
+    std::string FunctionCallerGenerator::getGlobalDeclarationsText() {
+        std::stringstream s;
+        for (auto &global : this->declarationManager->globals) {
+            s << "" << global.getType()->getCTypeName() << " " << global.getName() << ";" << std::endl;
+        }
+        return s.str();
+    }
+
+
+    std::string getJsonInputTemplateTextForJsonRvalue(Type &arg) {
+        std::stringstream output;
+        if (arg.isPointerTy())
+            output << getJsonInputTemplateTextForJsonRvalue(*arg.getPointerElementType()); // we abstract pointers away
+
+        // if scalar
+        if (arg.isArrayTy() == false && arg.isStructTy() == false && !arg.isPointerTy()) {
+            output << arg.getCTypeName();
+        }
+
+        if (arg.isArrayTy()) {
+            output << "[";
+            for (int i = 0; i < arg.getArrayNumElements(); i++) {
+                output << getJsonInputTemplateTextForJsonRvalue(*arg.getArrayElementType());
+                if (i != arg.getArrayNumElements() - 1) {
+                    output << ",";
+                }
+            }
+            output << "]";
+        }
+
+        if (arg.isStructTy()) {
+            output << "{";
+            for (auto &mem : arg.getNamedMembers()) {
+                output << "\"" << mem.getName() << "\"" << ": "
+                       << getJsonInputTemplateTextForJsonRvalue(*mem.getType());
+                if (mem.getName() != arg.getNamedMembers().back().getName()) { // names should be unique
+                    output << "," << std::endl;
+                }
+            }
+            output << "}";
+        }
+        return output.str();
+    }
+
+    void FunctionCallerGenerator::generate_json_input_template_file(std::string dest_file_path) {
+        std::ofstream of;
+        of.open(dest_file_path, std::ofstream::out | std::ofstream::trunc);
```

Comment:
Prefer direct initialization wherever possible:

```suggestion
        std::ofstream of(dest_file_path.c_str(), std::ofstream::out | std::ofstream::trunc);
```